packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -261,8 +261,10 @@ library LibProposing {
             }
 
             // Check that after hooks, the Taiko Token balance of this contract
-            // have increased by at least config.livenessBond
-            if (tko.balanceOf(address(this)) < tkoBalance + config.livenessBond) {
+            // have increased by the same amount as config.livenessBond (to prevent)
+            // multiple draining payments by a malicious proposer nesting the same
+            // hook.
+            if (tko.balanceOf(address(this)) != tkoBalance + config.livenessBond) {
```

Comment:
emm I guess another reason why we use `<` here is to allow hook providers to charge extra `TKO`?

Reply:
It is dangerous to handle like that. Hooks -imo - are not like 3rd party "providers" but more just an external libraries, extending the functionality.

Scneario:
1. Proposer gets a prover's signature to prove a block.
2. A malicious proposer can embedded that data into multiple hooks (per same` proposeBlock()`) which will drain the TKO balance - as liveness bond.
So basically this is to guarantee that the proposer does not overpay by embedding ("replaying") his/her signed assigment per same hook array.

Reply:
I would say to just limit the protocol to a single hook. If a prover/proposer wants to do multiple function calls that's fine, but they'll have to create a new hook smart contract to do that. Seems like a niche use case that we don't see people using directly, and if they want to do something like that they still can, but we keep the complexity outside of the main protocol.

With data blobs now (and even more efficient data copies from memory coming when still using calldata) it'll be more flexible and less complex to move custom logic on top of the propose function instead of inside it as well, so we'll be able to move more stuff out in future versions.

Reply:
Ok, verdict:
- I leave the API as is, but restrict to only call 1 hook. (revert if more)
- Also leave the equal here, as if the hook provider charges more, it will keep the TKO, and not forwards it to TaikoL1.

Reply:
> It is dangerous to handle like that. Hooks -imo - are not like 3rd party "providers" but more just an external libraries, extending the functionality.
> 
> Scneario:
> 
> 1. Proposer gets a prover's signature to prove a block.
> 2. A malicious proposer can embedded that data into multiple hooks (per same` proposeBlock()`) which will drain the TKO balance - as liveness bond.
>    So basically this is to guarantee that the proposer does not overpay by embedding ("replaying") his/her signed assigment per same hook array.

our own hook is simply one implementation, we don't even check if the hook used is our impl or another one from a third party.